curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1436,38 +1426,55 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force,
     std::map<uint64_t, DataCachePtr> tmp;
     curve::common::LockGuard lg(flushMtx_);
     CURVEFS_ERROR ret = CURVEFS_ERROR::OK;
-    {
-        WriteLockGuard writeLockGuard(rwLockChunk_);
-        tmp = dataWCacheMap_;
-    }
-    auto iter = tmp.begin();
-    VLOG(6) << "ChunkCacheManager start , size: " << tmp.size();
-    for (; iter != tmp.end(); iter++) {
-        VLOG(9) << "Flush datacache chunkPos:" << iter->second->GetChunkPos()
-                << ",len:" << iter->second->GetLen() << ",inodeId:" << inodeId
-                << ",chunkIndex:" << index_;
-        assert(iter->second->IsDirty());
-        ret = iter->second->Flush(inodeId, force, toS3);
-        if ((ret != CURVEFS_ERROR::OK) && (ret != CURVEFS_ERROR::NOFLUSH)) {
-            LOG(WARNING) << "dataCache flush failed. ret:" << ret
-                         << ",index:" << index_
-                         << ",data chunkpos:" << iter->second->GetChunkPos();
-            return ret;
+    DataCachePtr dataCache;
+    while (1) {
+        bool isFlush = false;
+        {
+            WriteLockGuard writeLockGuard(rwLockChunk_);
+
+            auto iter = dataWCacheMap_.begin();
+            while (iter != dataWCacheMap_.end()) {
+                if (iter->second->CanFlush(force)) {
+                    dataCache = iter->second;
+                    dataWCacheMap_.erase(iter);
+                    isFlush = true;
+                    break;
+                } else {
+                    iter++;
+                }
+            }
         }
-        WriteLockGuard lockGuard(rwLockChunk_);
-        if (ret == CURVEFS_ERROR::OK) {
-            if (!iter->second->IsDirty()) {
+        if (isFlush) {
+            VLOG(9) << "Flush datacache chunkPos:" << dataCache->GetChunkPos()
+                    << ",len:" << dataCache->GetLen() << ",inodeId:" << inodeId
+                    << ",chunkIndex:" << index_;
+            assert(dataCache->IsDirty());
+            do {
+                ret = dataCache->Flush(inodeId, toS3);
+                if (ret == CURVEFS_ERROR::NOTEXIST) {
+                    LOG(WARNING)
+                        << "dataCache flush failed. ret:" << ret
+                        << ",index:" << index_
+                        << ",data chunkpos:" << dataCache->GetChunkPos();
+                    ReleaseWriteDataCache(dataCache);
+                    break;
+                } else if (ret == CURVEFS_ERROR::INTERNAL) {
+                    ::sleep(3);
```

Comment:
why hard code 3 here? should it be configurable?

Reply:
add some log here ?

Reply:
> why hard code 3 here? should it be configurable?
CURVEFS_ERROR::INTERNAL  this scenario will only occur in a few cases, there is no need to retry a configuration separately

Reply:
> add some log here ?

fixed